WMF Framework/Event Platform Client/EPCNetworkManager.swift

```diff
@@ -0,0 +1,135 @@
+     */
+    private var outputBuffer: [(url: URL, body: NSDictionary)] = []
+    
+    init(session: Session = Session.shared) {
+        self.session = session
+    }
+
+    func schedulePost(url: URL, body: NSDictionary) {
+        appendPostToOutputBuffer((url: url, body: body))
+    }
+    
+    func httpDownload(url: URL, completion: @escaping (Data?) -> Void) {
+        httpDownload(url: url, retriesRemaining: 5, attemptDelay: 2, completion: completion)
+    }
+
+    /**
+     * Download data over HTTP with capacity for retries
+     * - Parameters:
+     *   - url: where to request data from
+     *   - retriesRemaining: maximum number of retries allowed for this download
+     *     operation
+     *   - attemptDelay: time between each retry
+     *   - completion: what to do with the downloaded data
+     *
+     * The operation will be performed up to `retriesRemaining + 1` times.
+     */
+    private func httpDownload(url: URL, retriesRemaining: Int, attemptDelay: TimeInterval, completion: @escaping (Data?) -> Void) {
+
+        if retriesRemaining < 0 {
+            completion(nil)
+            return
+        }
+        
+        let task = session.dataTask(with: url) { (data, response, error) in
+            
+            let failureBlock = {
+                dispatchOnMainQueueAfterDelayInSeconds(attemptDelay) {
+                    self.httpDownload(url: url, retriesRemaining: retriesRemaining - 1, attemptDelay: attemptDelay, completion: completion)
+                }
+            }
+            
+            guard let httpResponse = response as? HTTPURLResponse,
+                let data = data else {
+                    DDLogWarn("EPCNetworkManager: server did not respond adequately, will try \(url.absoluteString) again")
+                failureBlock()
+                return
+            }
+            
+            guard httpResponse.statusCode == 200 else {
+                DDLogWarn("EPCNetworkManager: HTTP status of response was \(httpResponse.statusCode), will try \(url.absoluteString) again")
+                failureBlock()
+                return
+            }
+            
+            guard error == nil else {
+                failureBlock()
+                return
+            }
+            
+            completion(data)
+        }
+        task?.resume()
+    }
+    
+    func httpTryPost() {
+        var queuedEvent: (url: URL, body: NSDictionary)?
+        while !outputBufferIsEmpty() {
+            queuedEvent = removeOutputBufferAtIndex(0)
+            if let queuedEvent = queuedEvent {
+                self.submit(url: queuedEvent.url, payload: queuedEvent.body)
+            }
+        }
+    }
+
+    private func submit(url: URL, payload: Any? = nil) {
+        let request = session.request(with: url, method: .post, bodyParameters: payload, bodyEncoding: .json)
+        let task = session.dataTask(with: request, completionHandler: { (_, response, error) in
```

Comment:
Just curious, is there an easy way to batch post multiple events with one call? I thought I remember some mention of that being possible with the endpoint. It would be an quick win if so and you could avoid the `while` in `httpTryPost`.

Reply:
Yep! We could definitely batch post multiple events with one call. From [T239996#5723920](https://phabricator.wikimedia.org/T239996#5723920):

> So with current settings, each message must be smaller than 1MB, and the entire POST body must be smaller than 4MB.

A conservative batch size would probably be 32 or 64 events. @tonisevener @linehan @mdholloway what do you think? Should we try batching in v1 or table it for next iteration and leave it one-at-a-time for now?

Reply:
There are arguments both ways. I personally feel like batching is fine, but you can make the argument that if the network (or server) is unreliable, batching risks losing more data if the unreliability is uniform (which is probably isn't, so in practice there is maybe less of a difference between these two in terms of expected data loss). 

Setting the moral arguments to one side, and though I want to get to it, maybe it makes sense to wait until next rev. We don't actually have anything using the batching functionality yet that I know of, so there may be unknown unknowns there. Maybe we can revisit if everything else is locked down and we have time for a little extra work to prove it out before merge, though.

Insert usual warning sticker that `while` loops are the leading cause of non-halting programs, etc. etc.

Reply:
👍  sounds good!

Reply:
Captured in wishlist for v2 ([T261987](https://phabricator.wikimedia.org/T261987))